feat(metrics): log flowEvents of all requests

seanmonstar opened the pull request:
This logs responses as flow events for routes we've determined are definitely part of a user flow. The format is `route.${path}.${status}`, such as `route./account/login.200` or `route./account/login.400.103`. Includes documentation.

cc @philbooth 
Closes #1526

rfk commented:
> we should maybe just build a whitelist of routes that should always log

This sounds OK to me at first glance, and I think we're better off optimizing for shipping it quickly in the routes we know we want.  So 👍

seanmonstar commented:
Updated to use a whitelist, which I built by scanning the `routes.js` file and adding routes where I also saw `request.emitMetricsEvent` calls. Also updated docs.

philbooth commented on docs/metrics-events.md:
```diff
@@ -67,6 +67,8 @@ in a sign-in or sign-up flow:
 |`password.forgot.resend_code.completed`|A password reset email has been re-sent to the user.|
 |`password.forgot.verify_code.start`|A user has clicked on the link in a password reset email.|
 |`password.forgot.verify_code.completed`|A password reset has been successfully completed on the server.|
+|`route.${path}.200`| A route responded with a 200 status code. Example: `route./account/login.200`|
```
For people consuming these events in redash, I wonder whether the `/` characters might seem slightly weird. Given that we've settled on `.` as a universal separator of sorts, do you think it is worth us converting these event names to something like `route.account.login.200`?

seanmonstar replied:
I initially suggested we replace the `/` with `.`, but then after building the signup funnel, I felt seeing the `/` would make it easier to recognize which route was reported. And staying slashes, we could build queries like `WHERE type LIKE 'route.%.4%` to grab requests that had 4XX responses (I think, I don't recall if MySQL's `%` is too greedy).

But if people have other objective reasons, I can switch!

philbooth commented on lib/log.js:
```diff
@@ -246,6 +249,32 @@ function optionallySetService (data, request) {
+  if (!FLOW_EVENT_ROUTES.has(path)) {
+    return
+  }
+  const code = response.statusCode || response.output.statusCode
```
What's the deal with falling back to `response.output` here? Does it warrant an explanatory comment?

seanmonstar replied:
Ah, this line was just basically copied from the `summary` method. I believe the fallback is that if `response instanceof Error`, it will have the status code on the output property.

philbooth commented:
@seanmonstar, I like this very much indeed, nice one!

Couple of inline comments but essentially I am r+. 👍

philbooth commented on docs/metrics-events.md:
```diff
@@ -67,6 +67,8 @@ in a sign-in or sign-up flow:
 |`password.forgot.verify_code.start`|A user has clicked on the link in a password reset email.|
 |`password.forgot.verify_code.completed`|A password reset has been successfully completed on the server.|
+|`route.${path}.200`| A route responded with a 200 status code. Example: `route./account/login.200`|
+|`route.${path}.${statusCode}.${errno}`| A route responded with a >=400 status code. Includes `errno`. Example: `route./account/login.400.103`|
```
Just noticed something else while editing this doc. Could you move this line down to the next table, which is specifically there to document error events?

rfk commented:
IIUC, this should have shipped in train-75, but I can't see any of the flow events in re:dash.  Do we need to do something special to enable them?

rfk commented:
I'm still not seeing this events in re:dash.  Is it possible they're being filtered out because of the non-alphanumeric event name?  /cc @philbooth @seanmonstar
